### PR TITLE
add tests for ddp with activation check points

### PR DESCRIPTION
**Pull request by zhaojuanmao**
Summary: add two success cases and two failure cases for ddp with activation check points when grad_as_bucket_view = true and false

Test Plan: unit tests

Differential Revision: D26679895

### CODE REVIEW

File: test/distributed/test_c10d.py [hunk 1]
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
...
+        else:
+            model = self.CheckpointTwiceModule().cuda()
+        model = nn.parallel.DistributedDataParallel(model,
+                                                    bucket_cap_mb=1,
+                                                    gradient_as_bucket_view=use_bucket_view,
+                                                    device_ids=[self.rank],
+                                                    process_group=process_group,
+                                                    find_unused_parameters=find_unused_parameters)
+        input_tensor = torch.rand((64, 2000)).cuda()
+        input_tensor.requires_grad = True
```
Comment (rohan-varma) [inline test/distributed/test_c10d.py:4168]: nit: I think you can specify requires_grad=True in constructor of tensor itself.

File: test/distributed/test_c10d.py [hunk 2]
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
...
+    def test_ddp_checkpointing_once(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        process_group = c10d.ProcessGroupNCCL(store, self.rank, self.world_size)
+        for use_bucket_view in (True, False):
+            model = self._test_ddp_checkpointing(checkpoint_once=True,
+                                                 process_group=process_group,
+                                                 use_bucket_view=use_bucket_view)
+            norm = 0.0
+            for p in model.parameters():
+                assert p.grad is not None
```
Comment (rohan-varma) [inline test/distributed/test_c10d.py:4185]: Use `self.assertTrue` here?

File: test/distributed/test_c10d.py [hunk 3]
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
...
+                norm += p.grad.norm().item()
+            assert numpy.allclose(norm, 78053.52966308594), norm
```
Comment (rohan-varma) [inline test/distributed/test_c10d.py:4187]: What does `assert numpy.allclose(norm, 78053.52966308594), norm` check for? Also, could we analytically compute the norm `78053` instead of hardcoding it?

File: test/distributed/test_c10d.py [hunk 4]
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
...
+                                                        gradient_as_bucket_view=use_bucket_view,
+                                                        device_ids=[self.rank],
+                                                        process_group=process_group)
+            input_tensor = torch.rand((64, 2000)).cuda()
+            input_tensor.requires_grad = True
+            output_tensor = checkpoint(model, input_tensor)
+            output_tensor.sum().backward()
+            norm = 0.0
+            for p in model.parameters():
+                assert p.grad is not None
```
Comment (rohan-varma) [inline test/distributed/test_c10d.py:4245]: same as above

File: test/distributed/test_c10d.py [hunk 5]
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
...
+                norm += p.grad.norm().item()
+            assert numpy.allclose(norm, 57004.33825683594), norm
```
Comment (rohan-varma) [inline test/distributed/test_c10d.py:4247]: same as above

File: torch/lib/c10d/reducer.cpp [hunk 6]
```diff
@@ -1190,20 +1190,12 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
...
-                    "multiple parameters sharing the same gradient. For "
-                    "example, param0 and param1 share the same gradient "
-                    "grad0. In this case, grad0 would first point to "
-                    "bucket_view_in0 when param0 is ready. Later, when "
-                    "param1 is ready, it will override grad0 to point to "
-                    "bucket_view_in1. However, param0 still expects grad0 "
-                    "to point to bucket_view_in0, and hence hit this "
-                    "warning. If you saw this message, please double-check if "
-                    "the above situation is expected for your application.");
+                    "expected DDP bucket view with gradient_as_bucket_view=True. "
```
Comment (rohan-varma) [inline torch/lib/c10d/reducer.cpp:1195]: not from this PR, but above line should end with " _where_ gradient is not the ..."

**Review by rohan-varma — APPROVED** [review]
Wonder if it's worth it to also have a more general deduplication check (unittest)? This might help us eliminate duplication bugs if the possibility comes up in another bug. 

Something like, have a model with duplicate parameters, create `Reducer`, and expose methods to access its parameters and ensure that they are deduped.

File: test/distributed/test_c10d.py (hunk 1 above)
Comment (mcarilli) [reply test/distributed/test_c10d.py:4168]: Cuda too, and we need to make sure not a record a copy to the GPU in the autograd history. Use the one-liner
```python
input_tensor = torch.rand((64, 2000), device="cuda", requires_grad=True)
```

File: test/distributed/test_c10d.py [hunk 7]
```diff
@@ -4082,6 +4084,168 @@ def allreduce_hook_gloo(
...
+            l2 = nn.Linear(2000, 2000)
+            l1.weight = l2.weight
+            model = nn.Sequential(l1, l2).cuda()
+            model = nn.parallel.DistributedDataParallel(model,
+                                                        bucket_cap_mb=1,
+                                                        gradient_as_bucket_view=use_bucket_view,
+                                                        device_ids=[self.rank],
+                                                        process_group=process_group)
+            input_tensor = torch.rand((64, 2000)).cuda()
+            input_tensor.requires_grad = True
```
Comment (mcarilli) [inline test/distributed/test_c10d.py:4240]:
```python
input_tensor = torch.rand((64, 2000), device="cuda", requires_grad=True)
```

File: test/distributed/test_c10d.py (hunk 3 above)
Comment (zhaojuanmao) [reply test/distributed/test_c10d.py:4187]: did not know how to calculate the norm..., do you have reference? the test should be good as we set the same seed "torch.manual_seed(31415)" to make sure results are reproducible.

**Comment by zhaojuanmao** [comment]
@rohan-varma regarding deduplication check (or unitest), I'm planning to revert the changes in  https://github.com/pytorch/pytorch/pull/46755 as right now we deduplicated params. along with it, I will add an assert if it is found that duplicated parameters sharing the same grad accumulator in the gradAccToVariableMap_ . ==> that should be sufficient for deduplication check

**Comment by mcarilli** [comment]
Do not merge yet, https://github.com/pytorch/pytorch/pull/51929/ got temporarily reverted (see https://github.com/pytorch/pytorch/pull/51929/#issuecomment-790271662). Give us a while to track down source of asan failures.

**Review by mcarilli — COMMENTED** [review]
"requesting changes" as a reminder not to land until we investigate why https://github.com/pytorch/pytorch/pull/51929/#issuecomment-790271662 got temporarily reverted (see the last comment on that PR, for some reason linking directly to it doesn't work).

**Review by mcarilli — CHANGES_REQUESTED** [review]
see https://github.com/pytorch/pytorch/pull/51929/#issuecomment-790271662
